torch/csrc/autograd/record_function_ops.cpp

```diff
@@ -29,13 +30,18 @@ at::Tensor record_function_enter(const std::string& name) {
   return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
 }
 
+RecordFunction& getRecordFunctionFromTensor(const at::Tensor& handle) {
+  auto& rec = at::cpp_custom_type_hack::cast<RecordFunction>(handle);
+  return rec;
+}
+
 void record_function_exit(const at::Tensor& handle) {
   // We don't actually need to do anything with handle just need to persist the
   // lifetime until now.
-  auto& rec = at::cpp_custom_type_hack::cast<RecordFunction>(handle);
+  auto& rec = getRecordFunctionFromTensor(handle);
   if (auto* current = RecordFunction::current()) {
-    AT_ASSERT(
-        current->name() == StringView("profiler::_record_function_exit"));
+    AT_ASSERT(current->name() == StringView("profiler::_record_function_exit"));
+    current->end();
```

Comment:
This seems like it was a bug from before, we would want `_record_function_exit` op to stop being profiled before the overall record_function scope, not after.

Reply:
isn't it the case of TorchScript, when this _record_function_exit op will be executed as a c10 op within TS and thus profiled?

Reply:
@ilia-cher It is automatically profiled, though shouldn't we end it early like we do for `record_function_start`? Otherwise, `profiler::_record_function_exit` will end after `rec`, which here is the `RecordFunction` corresponding to the user-created scope. The order should be:

push _record_function_start
pop _record_function_start
push rec
push _record_function_exit
pop record_function_exit
pop rec

Currently, the last 2 seem inverted to me, which this line aims to fix.

Reply:
I see, sg